utilities/measurement_utility.cc

```diff
@@ -15,95 +15,127 @@
 
 // make_measurement.exe --type=hash --input=input-file --output=output-file
 
+#include <sstream>
 #include <gflags/gflags.h>
 #include "certifier.h"
 #include "support.h"
 
+using namespace std;
 using namespace certifier::utilities;
 
+int parse_other_files_size(string &        other_files,
+                           vector<string> &other_files_list,
+                           vector<int> &   other_files_size);
+
 DEFINE_bool(print_all, false, "verbose");
 DEFINE_bool(print_debug, false, "print debugging info");
 DEFINE_string(type, "hash", "measurement type");
 DEFINE_string(input, "measurement_utility.exe", "input file");
-DEFINE_string(policy_key, "", "policy key file, e.g., policy_key.py");
+DEFINE_string(other_files,
+              "",
+              "Comma-separated list of other files to include in measurement; "
+              "e.g., policy_key.py,certifier_framework.py");
 DEFINE_string(output, "measurement_utility.exe.measurement", "output file");
 
 
 const int sha256_size = 32;
 
-int hash_utility(string &input, string &policy_key_file, string &output) {
+// other_files is a comma-separated list of other-files to include in the
+// measurement
+int hash_utility(string &input, string &other_files, string &output) {
 
   int in_size = file_size(input);
   int in_read = in_size;
   int to_hash_size = in_size;
 
-  int policy_key_size =
-      (policy_key_file.size() ? file_size(policy_key_file) : 0);
+  vector<string> other_files_list;
+  vector<int>    other_files_size;
 
-  to_hash_size += policy_key_size;
+  // (other_files.size() ? file_size(other_files) : 0);
+  int other_files_total_size =
+      (other_files.size() ? parse_other_files_size(other_files,
+                                                   other_files_list,
+                                                   other_files_size)
+                          : 0);
 
-  byte *       to_hash = (byte *)malloc(to_hash_size * sizeof(byte) + 1);
+  if (other_files_total_size < 0) {
+    printf("Error, reading one or more input files.\n");
+    return 1;
+  }
+  to_hash_size += (other_files_total_size + 1);
+
+  byte *       to_hash = (byte *)malloc(to_hash_size * sizeof(byte));
```

Comment:
you don't need to load all files in memory to compute this hash -- you can load files one by one and update the hash

Reply:
If you load them one by one there is a TOCTOU vulnerability, I think.

Reply:
Although this is not a big deal in a utility

Reply:
Hi, I went back and did some investigation on @rgerganov 's comment -- that would have simplified the code as we would not need to allocate a huge buffer to read-in contents of multiple files.

However, @jlmucb 's comment is helpful -- so I retained the code as it was submitted for review.

Still, before I got John's comment, I experimented whether Rado's suggestion would result in the same hash for multiple files, v/s running `digest_message()` across each file, and have it update the hash.

I maybe misunderstanding this approach, but take a look at the new test case added in `support_tests.cc/test_digest_multiple()` See my comments in that test case, which seems to indicate that as coded both these schemes are not similar. They result in different hash-digest values.

Reply:
I don't think that hashing files one by one creates any vulnerability here